### DISCUSSION TITLE
Stop ImmersiveTranslate updating translation when the source text changes

### DISCUSSION THREAD

**Discussion by iG8R** [immersive-translate/immersive-translate#1598] [Help]
### 概述 | Summary

Sometimes, when using Gemini, I encounter situations when some untranslated words or phrases remain in the translated text.
In such cases, I try using FoxReplace addon (https://addons.mozilla.org/en-US/firefox/addon/foxreplace/) to replace these leftovers with reasonable replacements. But when I do this and changes also occur in the original text, IT always tries to re-translate the page.
Is there a way to prevent IT from tracking changes in already translated original texts?

E.g. URL - https://www.69shu.pro/txt/13823/8196298

<img src="https://github.com/immersive-translate/immersive-translate/assets/11407417/41d38dff-12f9-40c9-835f-93880a5873c5" width="550">
<img src="https://github.com/immersive-translate/immersive-translate/assets/11407417/0be887cc-1db9-492f-8855-f6d6e9a12b74" width="550">
<img src="https://github.com/immersive-translate/immersive-translate/assets/11407417/7ccf2570-ae7d-4425-a6a9-9aacb946fef1" width="550">
<img src="https://github.com/immersive-translate/immersive-translate/assets/11407417/f22e6473-d852-4f83-8293-c09e0a2dbbb1" width="550">

[ImmersiveTranslate-FoxReplace-Video_2024-04-03_11-44-02.webm](https://github.com/immersive-translate/immersive-translate/assets/11407417/300d73a7-6557-494a-a313-1c8dafd3dbaf)

PS. I can't disable JavaScript after IT has done its initial work, because FoxReplace also uses JavaScript to replace necessary words.

PPS. And "yes", I've already tried all possible prompts with Gemini to make it not to leave any untranslated words or phrases.
For example:
```
    "gemini": {
      "APIKEY": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
      "limit": "1",
      "maxTextGroupLengthPerRequest": "15",
      "model": "gemini-pro",
      "prompt": "user: You are a professional translation engine, please translate the text into a colloquial, professional, elegant and fluent content, without the style of machine translation. You must only translate the text content, never interpret it. In your answer never add anything except the translated text. The entire text should be translated, in the translated text do not leave a single word or phrase untranslated or in the original language.\n\nmodel: Ok, I will only translate the text content, never interpret it.\n\nuser: Translate into English\n你好\n世界\n\nmodel:\nHello\nWorld\n\nuser: Translate  into {{to}}\n{{text}}",
      "apiUrl": "https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={key}",
      "maxTextLengthPerRequest": "2000"
    }
```

**Comment by keyboard3** [comment]
Because you changed the original text, if you don't want this situation to happen, you need to replace the content with the FoxReplace plugin before triggering the translation.

**Comment by iG8R** [comment]
That's the point, if I make changes to the original text before translation by IT, it will result in a mix of Chinese and English in the original text, which will lead to a sharp deterioration in the translation.

What I need is for IT to no longer react to changes in the original text after it has finished translating, trying to re-translate the entire page.

How can this be achieved without disabling all scripts on the page?
Maybe there could be an option that controls whether ImmersiveTranslate responds to changes in already translated text (rather than new content added by page scripts)?

**Comment by KnIfER** [comment]
替换时把所有 mutationobserver  关掉即可。 需要修改 conent_script.js，劫持所有mutationobserver   的构造函数。约定当窗口内含有id为x的元素时，不触发 mutationobserver  的回调函数。
